cli/src/index.js

```diff
@@ -0,0 +1,334 @@
+  tracker.progressOffset = 0;
+  tracker.totalOffset = 0;
+  let status = "";
+  tracker.setStatus = text => {
+    status = text || "";
+    update();
+  };
+  let progress = "";
+  tracker.setProgress = (done, total) => {
+    spinner.prefixText = kleur.dim(`${done}/${total}`);
+    const completeness =
+      (tracker.progressOffset + done) / (tracker.totalOffset + total);
+    progress = kleur.cyan(
+      `▐${"▨".repeat((completeness * 10) | 0).padEnd(10, "╌")}▌ `
+    );
+    update();
+  };
+  function update() {
+    spinner.text = progress + kleur.bold(status) + getResultsText();
+  }
+  tracker.finish = text => {
+    spinner.succeed(kleur.bold(text) + getResultsText());
+  };
+  function getResultsText() {
+    let out = "";
+    for (const [filename, result] of results.entries()) {
+      out += `\n ${kleur.cyan(filename)}: ${prettyPrintSize(result.size)}`;
+      for (const { outputFile, outputSize, infoText } of result.outputs) {
+        const name = (program.suffix + extname(outputFile)).padEnd(5);
+        out += `\n  ${kleur.dim("└")} ${kleur.cyan(name)} → ${prettyPrintSize(
+          outputSize
+        )}`;
+        const percent = ((outputSize / result.size) * 100).toPrecision(3);
+        out += ` (${kleur[outputSize > result.size ? "red" : "green"](
+          percent + "%"
+        )})`;
+        if (infoText) out += kleur.yellow(infoText);
+      }
+    }
+    return out || "\n";
+  }
+  spinner.start();
+  return tracker;
+}
+
+async function processFiles(files) {
+  const parallelism = cpus().length;
+
+  const results = new Map();
+  const progress = progressTracker(results);
+
+  progress.setStatus("Decoding...");
+  progress.totalOffset = files.length;
+  progress.setProgress(0, files.length);
+
+  const workerPool = new WorkerPool(parallelism, __filename);
+  // Create output directory
+  await fsp.mkdir(program.outputDir, { recursive: true });
+
+  let decoded = 0;
+  let decodedFiles = await Promise.all(
+    files.map(async file => {
+      const result = await workerPool.dispatchJob({
+        operation: "decode",
+        file
+      });
+      results.set(file, {
+        file: result.file,
+        size: result.size,
+        outputs: []
+      });
+      progress.setProgress(++decoded, files.length);
+      return result;
+    })
+  );
+
+  for (const [preprocessorName, value] of Object.entries(preprocessors)) {
+    if (!program[preprocessorName]) {
+      continue;
+    }
+    const preprocessorParam = program[preprocessorName];
+    const preprocessorOptions = Object.assign(
+      {},
+      value.defaultOptions,
+      JSON5.parse(preprocessorParam)
+    );
+
+    decodedFiles = await Promise.all(
+      decodedFiles.map(async file => {
+        return workerPool.dispatchJob({
+          file,
+          operation: "preprocess",
+          preprocessorName,
+          options: preprocessorOptions
+        });
+      })
+    );
+
+    for (const { file, bitmap, size } of decodedFiles) {
```

Comment:
What's happening here? Can you add a comment in the body if this loop is necessary?

Reply:
lol wtf. That loop is not needed :D